### ISSUE TITLE
Unrestrict subscribe?

### ISSUE THREAD

**Issue by aneophyte** [airbnb/MvRx#85]
As it currently seems to be the case, only `Fragment`s have been implemented as `MvRxView`. It certainly does make sense for apps that have minimal number of `Activity`s and large number of `Fragment`s. I'm working on an app, it has a lot of `Activity`s, some of them do not even include any `Fragment`. To be able to use the `Activity`s as `MvRxView`s, with little effort, I managed to implement `MvRxView` for them. I followed the exact same way it was for `Fragment`s. I haven't done extensive testing but it works.

However, the problem in this is that I'm using a [function](https://github.com/airbnb/MvRx/blob/88e9ac22d4e0cab23acb5418cbd07a7acac06186/mvrx/src/main/kotlin/com/airbnb/mvrx/BaseMvRxViewModel.kt#L169) that is restricted to the library package. Would it be ok if it was made public?

**Comment by aneophyte** [comment]
This is what I did:
```kotlin
abstract class BaseActivity : BaseMvRxActivity(), MvRxView {
    override val mvrxViewModelStore by lazy { MvRxViewModelStore(viewModelStore) }

    override fun onCreate(savedInstanceState: Bundle?) {
        super.onCreate(savedInstanceState)
        mvrxViewModelStore.restoreViewModels(this, savedInstanceState)
    }

    override fun onSaveInstanceState(outState: Bundle) {
        super.onSaveInstanceState(outState)
        mvrxViewModelStore.saveViewModels(outState)
    }

    override fun onStart() {
        super.onStart()
        // This ensures that invalidate() is called for static screens that don't
        // subscribe to a ViewModel.
        postInvalidate()
    }
}

inline fun <T, reified VM : BaseMvRxViewModel<S>, reified S : MvRxState> T.activityViewModel(
        viewModelClass: KClass<VM> = VM::class,
        noinline keyFactory: () -> String = { viewModelClass.java.name }
) where T : AppCompatActivity, T : MvRxView = lifecycleAwareLazy(this) {
    val stateFactory: () -> S = { _activityViewModelInitialStateProvider(keyFactory) }
    if (this !is MvRxViewModelStoreOwner) throw IllegalArgumentException("Your Activity must be a MvRxViewModelStoreOwner!")
    MvRxViewModelProvider.get(viewModelClass.java, this, keyFactory(), stateFactory)
            .apply { subscribe(this@activityViewModel, subscriber = { postInvalidate() }) } //red line here
}

@Suppress("FunctionName")
inline fun <reified S : MvRxState, T : AppCompatActivity> T._activityViewModelInitialStateProvider(keyFactory: () -> String): S {
    val args: Any? = intent.getBundleExtra(MvRx.KEY_ARG)
    if (this is MvRxViewModelStoreOwner) {
        this.mvrxViewModelStore._saveActivityViewModelArgs(keyFactory(), args)
    } else {
        throw IllegalArgumentException("Your Activity must be a MvRxViewModelStoreOwner!")
    }
    return _initialStateProvider(S::class.java, args)
}
```

**Comment by BenSchwab** [comment]
@aneophyte You can use the subscribe methods exposed in `MvRxView`, so just delete `this@activityViewModel`.

However -- we haven't been consistent on this -- but all the methods named starting with `_` are also not meant to be publicly used.

I feel like a solution here is to change `MvRxExtensions.viewModel` to apply a subscribe call if the receiver is a `MvRxView`. 

However, I would be hesitant to make this change without first understanding a bit more about what your refactoring to MvRx looks like. One of MvRx opinion's is that activities are just shells for fragments. It makes sense that this is different than may exist in applications which want to switch to MvRx. However I would imagine you will already need to refactor your activity to confirm to the `invalidate` model, and am curious how much extra work it would be to pull it into a fragment?

**Comment by aneophyte** [comment]
I haven't used any subscribe method outside the extension function. The `subscribe` method in question came because it was part of the `activityViewModel` method. This is the method, I believe, that is used to create view model under the activity scope. It leads me to another question that if that subscribe method were to be removed, would persistence and lifecycle related stuff work?

`MvRxExtensions.viewModel` is definitely the better option, though the restriction is still there. I tried to make an extension function around it.

This is how I'm using `activityViewModel` currently.
```kotlin
class SimpleActivity : BaseActivity() {
    val viewModel by activityViewModel(ActivitySplashViewModel::class)

    override fun invalidate() {
        withState(viewModel) {
        }
    }

    override fun onCreate(savedInstanceState: Bundle?) {
        super.onCreate(savedInstanceState)
        setContentView(R.layout.activity_simple)
    }
}
```
kotlin
The problem with refactoring to `Fragment`s, here in my case, is a lot has been written already and would require quite a lot of changing. And moving from `Activity` based flow to `Fragment` one, I suppose would require using the Navigation library from the AndroidX family. I'm a bit hesitant with that because it's currently in experimental state, also I did not have pleasant experience in using `Fragment`s as a prominent thing in previous projects.

Having said all this, maybe applying your solution would allow `Activity`s to be used as `View`s, would make the library more open for developers opting to use `Activity`, but to be honest, I don't really want my requirements to affect the conventions that MvRx has opted for.

Thanks for the input! I apologize for responding late.
